grub2/install

```diff
@@ -7,6 +7,126 @@
 #   bootloader, language
 #
 
+# Note: if target is "i386-pc" and /etc/default/grub_installdevice is
+# missing or does not contain any valid device entry grub will be installed
+# on the disk that contains the /boot directory.
+#
+# If /boot happens to be on a RAID or similar which is composed of several
+# physical disks then grub is installed only on one (the 'first') of them.
+#
+
+# Usage: is_disk(disk)
+#
+# Return disk if disk is a disk device (not a partition), else "".
+# Disk name is without leading '/dev/'.
+#
+function is_disk
+{
+  local dev=$1
+
+  if [ -d "/sys/block/$dev" -a -b "/dev/$dev" ] ; then
+    echo $dev
+  fi
+}
+
+
+# Usage: partition_to_disk(partition)
+#
+# Find disk device name for a partition.
+# Partition and disk name are without leading '/dev/'.
+#
+# Return empty string if there is no parent disk device for a partition.
+#
+function partition_to_disk
+{
+  local part=$1
+  local dev=$(echo /sys/block/*/$part)
+  # there should normally be just one entry - but better make sure
+  dev=${dev%% *}
+
+  if [ -d "$dev" ] ; then
+    dev=${dev#/sys/block/}
+    dev=${dev%/$part}
+    [ -b "/dev/$dev" ] && echo $dev
+  fi
+}
+
+
+# Usage: first_slave(disk)
+#
+# Find first slave device for disk.
+#
+# Return empty string if there are no slaves.
+#
+function first_slave
+{
+  local dev=$(echo /sys/block/$1/slaves/*)
+  dev=${dev%% *}
+  dev=${dev##*/}
+
+  [ "$dev" = "*" ] && dev=
+
+  echo $dev
+}
+
+
+# Usage: disk_device(arg)
+#
+# Find disk device name for arg.
+# arg may be a partition name or a directory
```

Comment:
From reading the "df" piece at lines 89-95, IIUC the partition name must start with `/dev`, unlike the output. That could be a separate `function`, OTOH it is not necessary because we only pass in `/boot`. Delete?

Reply:
I'd like to keep it in its current form even though the function is only called with '/boot' as argument.